File: java/526_BeautifulArrangement.java
class Solution {
    int count = 0;
    public int countArrangement(int n) {
        boolean[] visited = new boolean[n + 1];
        return backtrack(n, 1, visited);
    }
    
    public int backtrack(int n, int idx, boolean[] visited) {
        if (idx > n) return ++count;
        for (int i = 1; i <= n; ++i) {
            if (!visited[i] && (i % idx == 0 || idx % i == 0)) {
                visited[i] = true;
                backtrack(n, idx + 1, visited);
                visited[i] = false;
            }
        }
        return count;
    }
}

File: python/526_BeautifulArrangement.py
class Solution:
    def countArrangement(self, n: int) -> int:
        def backtrack(idx):
            nonlocal count
            if idx > n:
                count += 1
                return count
            
            for i in range(1, n + 1):
                if not visited[i] and (idx % i == 0 or i % idx == 0):
                    visited[i] = True
                    backtrack(idx + 1)
                    visited[i] = False
            
            return count
        
        count = 0
        visited = [False for _ in range(n + 1)]
        return backtrack(1)
